File: 1218-lowest-common-ancestor-of-deepest-leaves/1218-lowest-common-ancestor-of-deepest-leaves.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
typedef struct TreeNode TreeNode;
typedef struct Pair {
    TreeNode* node;
    int depth;
} pair;

pair* init(TreeNode* root, int depth) {
    pair* p = (pair*)malloc(sizeof(pair));
    p->node = root;
    p->depth = depth;
    return p;
}

pair* dfs(TreeNode* root) {
    if (root == NULL)
        return init(NULL, 0);

    pair* left = dfs(root->left);
    pair* right = dfs(root->right);

    if (left->depth > right->depth)
        return init(left->node, left->depth + 1);

    if (left->depth < right->depth)
        return init(right->node, right->depth + 1);

    return init(root, left->depth + 1);
}
struct TreeNode* lcaDeepestLeaves(TreeNode* root) {
    return dfs(root)->node;
}

File: 1218-lowest-common-ancestor-of-deepest-leaves/1218-lowest-common-ancestor-of-deepest-leaves.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Pair {
    public TreeNode root;
    public int depth;
    public Pair (TreeNode root, int depth) {
        this.root = root;
        this.depth = depth;
    }
}

public class Solution {
    private Pair dfs (TreeNode root) {
        if (root == null)
            return new Pair(null, 0);
        
        Pair left = dfs (root.left);
        Pair right = dfs (root.right);
        
        if (left.depth > right.depth)
            return new Pair(left.root, left.depth + 1);
        
        if (left.depth < right.depth)
            return new Pair(right.root, right.depth + 1);
        
        return new Pair(root, left.depth + 1);    
    }
    
    public TreeNode LcaDeepestLeaves(TreeNode root) {
        return dfs(root).root;
    }
}